### PR TITLE
[wgsl] Support for composite constant expressions.

### PR DESCRIPTION
**Pull request by lachlansneff**
Adds support for composite constant expressions to the wgsl frontend, as well as fixes a bug with broken array template endings.

### CODE REVIEW

File: src/front/wgsl.rs [hunk 1]
```diff
@@ -456,6 +458,27 @@ impl Parser {
         }
     }
 
+    fn deconstruct_composite_type(
+        type_arena: &mut Arena<crate::Type>,
+        ty: Handle<crate::Type>,
+        index: usize,
+    ) -> Result<Handle<crate::Type>, Error<'static>> {
+        let ty = match &type_arena[ty].inner {
```
Comment (kvark) [inline src/front/wgsl.rs:466]: let's not match on a reference, please

Comment (lachlansneff) [reply src/front/wgsl.rs:466]: How should I do this then? If I match on the value, it'll try to move it and it can't.

Comment (kvark) [reply src/front/wgsl.rs:466]: Use `ref` in pattern matching instead, where the data is not `Copy`:
```rust
let ty = match type_arena[ty].inner {
 ...
  crate::TypeInner::Struct { ref members } => {..}
};
```

Comment (lachlansneff) [reply src/front/wgsl.rs:466]: There's some ownership issues with that, but I'll work around them. For future reference, why is matching on a reference discouraged?

Comment (kvark) [reply src/front/wgsl.rs]: `ref` is precisely what you get behind the sugar of matching on a reference, and that's why I don't like it.
Sorry, it could be one of my pet peeves. I'm used to Rust that doesn't even allow one to match on a reference, and you had to write out `ref` where applicable. That has the benefit of you knowing precisely what's going on there upon reading some code, and it allows you to use `ref` even on `Copy` types in case they are big, and you don't really want them to be copied anywhere.

TL;DR: there shouldn't be any more ownership issues with `ref` than you'd get with matching on a reference

Comment (lachlansneff) [reply src/front/wgsl.rs]: Weird, maybe I was doing it wrong, but it was borrowing the `type_arena` through the whole match expression when I need to use it mutably. Whereas, matching on references didn't do that.

Comment (kvark) [reply src/front/wgsl.rs]: if you need mutability, you do `ref mut` instead of `ref`

File: src/front/wgsl.rs [hunk 2]
```diff
@@ -466,6 +468,34 @@ impl Parser {
...
     }
 
+    fn deconstruct_composite_type(
+        type_arena: &mut Arena<crate::Type>,
+        ty: Handle<crate::Type>,
+        index: usize,
+    ) -> Result<Handle<crate::Type>, Error<'static>> {
+        let ty = match type_arena[ty].inner {
+            crate::TypeInner::Vector {
+                ref kind,
```
Comment (kvark) [inline src/front/wgsl.rs:478]: you don't need to take `kind` and `width` by reference, they are small and copyable

File: src/front/wgsl.rs [hunk 3]
```diff
@@ -466,6 +468,34 @@ impl Parser {
...
+            }
+            | crate::TypeInner::Matrix {
+                ref kind,
+                ref width,
+                ..
+            } => Err(crate::TypeInner::Scalar {
+                kind: *kind,
+                width: *width,
+            }),
+            crate::TypeInner::Array { ref base, .. } => Ok(*base),
```
Comment (kvark) [inline src/front/wgsl.rs:490]: same for `base`

File: src/front/wgsl.rs (hunk 2 above)
Comment (lachlansneff) [reply src/front/wgsl.rs]: Yeah, I attempted writing that match again, and it now works out normally with `ref`s and all that.
